### DISCUSSION TITLE
Welcome to console Discussions!

### DISCUSSION THREAD

**Discussion by theofidry** [theofidry/console#89] [General]
This discussion, like the other discussions' first message is often edited to better capture the evolution of the various discussions. The goal is to properly define various issues, why they are a problem and find remedies for them. This space is meant to be more flexible and noisy than `symfony/symfony` to foster easier collaboration before presenting potential solutions to the Symfony repository.

# Table of Content

Random ideas:

- ~Integrate the configuration aspect that Kevin worked on a lot~
- ~Breaking down Application with the Runner~
- Other command interfaces
- #111
- Typedhinted arguments/options/coercing process
- Facilitate testing
  - #217
  - #191
  - normalized output
  - exit code
- #218
- Have the reversibility of the argument/option parsing
- Have access to the provided (i.e. raw) args & options
- Idiomatic way to smoke test a command?

**Comment by kbond** [comment]
- ~commands as any invokable (closure, class, method) like controllers~

**Comment by theofidry** [comment]
I'll post my draft notes from the SfCon here; I wanted to rewrite them more properly but I think it's best to actually expand it into another meta post that we can enrich & rework. I'll publish a first draft of it soon.

<hr/>
# Symfony Console improvements review

Notes of the meeting at the SymfonyCon22 the 17/11/2022 between the following participants:

- Kevin Bond (@kbond)
- Robin Chalas (@chalasr)
- Théo Fidry (@theofidry)
- Titouan Galopin (@tgalopin)

## Integrate the attributes

The goal: align the command usage with the controllers.

A solution: have a command decorator, which can wrap the input definition to enrich the native one with the parsed data.
When returning the command, the Application can support several decorators to return the decorated command instead of the decorator. There is a precedent for this for example with the LazyCommand.

For handling service injection via attribute: could be handled by the FrameworkBundle application but not in the Console Application.


## Commands as any invokable (closure, class, method) like controllers

Could leverage the `#[AsCommand]` attribute.

## Support multiple commands per class

Probably not a good idea due to the interactive/initialize API: it would be confusing to which command the initialize command is applied to.


## About support of injecting services

It is done in the controllers per the idea is that because a controller can have multiple action methods which cannot be the case for the command.

Also would not make sense to extend it to interact/initialize.

With PHP 8.1 property augmentation, the boiler-template is also reduced.

Also makes testing easier because constructing an object ready to test is going to be easier.

## Breaking down Application with the Runner

We could introduce the FidryApplication as ApplicationConfiguration and do new Application(ApplicationConfiguration) or smth along those lines.


### IO object

IO could be made into a big helper:

- input
- output
- style interface
- logger

Could switch/make more explit stdout/stderr write

Should not put the IO object as part of the command state via a `setIO()` -> this would make calling another command weirder and less convenient.
